### ISSUE TITLE
JSCH based client and mod_sftp negotiation fail

### ISSUE THREAD

**Issue by eeboaz** [proftpd/proftpd#830]
We have an internally developed (JSch-based, http://www.jcraft.com/jsch/) embedded SFTP that we use for getting diagnostic data to a central repository. When we performed updates on the host OS (CentOS 7-based), which included a newer version of proftpd and mod_sftp we started getting consistent negotiation failures from some of the deployed clients, but not all.

If we force the a test client to aes128-ctr, then the data transfer works, but we do need to use the aes256-ctr cipher and this worked across all clients before the upgrade to proftpd-1.3.5e.

Since the issue affected existing systems only after we did the upgrade on the host OS, the problem of negotiating aes256-ctr to be related to mod_sftp or a possible configuration issue.

I've attached trace logs for review - thank you.

proftpd -V output: https://pastebin.com/raw/uJ3aaA4J
proftpd configuration: https://pastebin.com/raw/DgjVDC08
[mod_sftp_jsch_negotiation_failure.txt](https://github.com/proftpd/proftpd/files/3601793/mod_sftp_jsch_negotiation_failure.txt)

**Comment by Castaglia** [comment]
Which version of OpenSSL are you using?  That is, what does `openssl version -a` show?

**Comment by Castaglia** [comment]
Here, in the logs, we see the issue:
```
2019-09-11 15:22:56,929 [30800] <ssh2:13>: received expected SSH_MSG_NEWKEYS message
2019-09-11 15:22:56,929 [30800] <ssh2:9>: setting session keys
2019-09-11 15:22:56,929 [30800] <ssh2:19>: waiting for max of 600 secs while polling socket 0 using select(2)
2019-09-11 15:22:56,937 [30800] <ssh2:20>: SSH2 packet len = 581289507 bytes
2019-09-11 15:22:56,937 [30800] <ssh2:20>: SSH2 packet padding len = 162 bytes
2019-09-11 15:22:56,937 [30800] <ssh2:20>: SSH2 packet payload len = 581289344 bytes
2019-09-11 15:22:56,937 [30800] <ssh2:20>: payload len (581289344 bytes) exceeds max payload len (262144), ignoring payload
```
The `NEWKEYS` SSH2 request puts the negotiated ciphers, keys into effect.  After that, we see a packet encoding error -- the encoded packet length is far too long.

Are you still able to reproduce/experience this issue when using a newer ProFTPD version, _e.g._ 1.3.7rc2 or 1.3.6b?

**Comment by eeboaz** [comment]
The version of OpenSSL is 1.0.2k (CentOS 7.x) and I wouldn't be able to test against a newer version of the server software because of internal policies.

After a few weeks of internal discussion, I was able to get access to the client source and found this was actually a client side issue, which I patched last week and confirmed last week and forgot to update this issue. The client never properly implemented cipher negotiations because the original author didn't understand the JSch framework. It was amazing it 'worked' in the first place, much less as long as it did.

The odd thing is that the file we were testing with was only a few MiB in size, nowhere near the size specified in the payload length message. 

This wasn't a ProFTPd/mod_sftp issue and thank you for your time, I'll go ahead and close this.
